libvast/src/systemd.cpp

```diff
@@ -0,0 +1,48 @@
+#include "vast/systemd.hpp"
+
+#include "vast/detail/posix.hpp"
+#include "vast/error.hpp"
+#include "vast/logger.hpp"
+
+#include <caf/detail/scope_guard.hpp>
+
+#include <cstdlib>
+#include <cstring>
+
+#include <sys/socket.h>
+
+namespace vast::systemd {
+
+#if VAST_LINUX
+
+// This function implements the `sd_notify()` protocol to signal readyness
+// to the service manager (systemd). This code follows the behaviour of the
+// reference implementation at `libsystemd/sd-daemon/sd-daemon.c` to decide
+// which conditions should result in errors.
+caf::error notify_ready() {
+  caf::detail::scope_guard guard([] {
+    ::unsetenv("NOTIFY_SOCKET"); // Always unset $NOTIFY_SOCKET.
+  });
+  auto notify_socket_env = ::getenv("NOTIFY_SOCKET");
+  if (!notify_socket_env)
+    return caf::none;
+  VAST_VERBOSE_ANON("notifying systemd at", notify_socket_env);
+  int socket = ::socket(AF_UNIX, SOCK_DGRAM | SOCK_CLOEXEC, 0);
+  if (socket < 0)
+    return make_error(ec::system_error, "failed to create unix socket");
+  if (detail::uds_sendmsg(socket, notify_socket_env, "READY=1\n", MSG_NOSIGNAL)
+      < 0)
```

Comment:
Should this use the `unix_domain_socket` wrapper that we have instead such that the user (i.e., the highlighted code) does not need to know about the return values?

Reply:
I don't think so; the `unix_domain_socket` seems to be purpose-built to send and receive file descriptors. Adding `sendmsg()` but not e.g. `send()` or `recvmsg()` as well would make for an odd API.